packages/myst-cli/src/process/search.ts

```diff
@@ -0,0 +1,152 @@
+import type { GenericNode } from 'myst-common';
+import type { DocumentHierarchy } from 'myst-spec-ext';
+import type { Heading } from 'myst-spec';
+import { toText } from 'myst-common';
+export const SKIP = () => {};
+
+type Visit = (content: GenericNode, path: string) => typeof SKIP | void;
+type Depart = (content: GenericNode, path: string) => void;
+
+function visitNode(content: GenericNode, visit: Visit, depart: Depart | undefined, path: string) {
+  const result = visit(content, path);
+  if (result !== undefined && result === SKIP) {
+    return;
+  }
+  if ('children' in content && content.children) {
+    visitNodeArray(content.children, visit, depart, `${path}.`);
+  }
+
+  depart?.(content, path);
+}
+
+function visitNodeArray(
+  content: GenericNode[],
+  visit: Visit,
+  depart: Depart | undefined,
+  path: string,
+) {
+  content.forEach((n, i) => {
+    visitNode(n, visit, depart, `${path}${i}`);
+  });
+}
+
+export function walk(content: GenericNode, visit: Visit, depart?: Depart, basePath?: string) {
+  const path = basePath ?? '$';
+  if (Array.isArray(content)) {
+    visitNodeArray(content, visit, depart, path);
+  } else {
+    visitNode(content, visit, depart, path);
+  }
+}
+
+export type HeadingInfo = {
+  text: string;
+  depth: number;
+  html_id?: string;
+};
+
+export type Section = {
+  heading?: HeadingInfo;
+  parts: string[];
+};
+
+export function toSectionedParts(content: GenericNode) {
```

Comment:
This naively sections a document upon encountering a heading. Seems to work well so far.